chess/board.h

```diff
@@ -26,21 +33,31 @@ namespace space {
 		Move(int v_sourceRank, int v_sourceFile, int v_destinationRank, int v_destinationFile, PieceType v_promotedPiece = PieceType::None) :
 			sourceRank(v_sourceRank), sourceFile(v_sourceFile), destinationRank(v_destinationRank), destinationFile(v_destinationFile), promotedPiece(v_promotedPiece)
 		{}
-		inline std::string as_string() const {
+		Move(const std::string &s);
+
+		std::string toString() const;
+		bool operator <(const Move& that) const;
+
+        inline std::string as_string() const {
 			std::stringstream ss;
 			ss << (char)('a' + sourceFile) << (sourceRank + 1)
 			   << " -> "
 			   << (char)('a' + destinationFile) << (destinationRank + 1);
 			return ss.str();
 		}
-		bool operator <(const Move& that) const;
 	};
+  
 	struct Piece {
 		PieceType pieceType;
 		Color color;
-		char as_char() const;
-	};
-	class IBoard {
+		Piece() {}
+		Piece(PieceType _p, Color _c) : pieceType(_p), color(_c) {}
+		Piece(char c);
```

Comment:
Umm... By adding these constructors, you are adding a performance penalty, since this is no longer a plain old data type.
Consider removing these constructors.
https://en.cppreference.com/w/cpp/named_req/PODType